File: src/core/resposne/response.ts
// interface Response<T> {
//     readonly result: string,
//     readonly message?: string
//     readonly data?: T,
// }

export interface Success<T> {
    readonly data: T
}

export interface Failure {
    readonly message: string
}

export type Response<T> = Success<T> | Failure

export function isSuccess<T>(response: Response<T>): response is Success<T> {
    return "data" in response;
}

File: src-tauri/src/commands/response.rs
use serde::Serialize;

#[derive(Serialize)]
#[serde(untagged)]
pub enum Response<T> {
    Success { result: String, data: T },
    Failure { result: String, message: String },
}

impl<T> Response<T> {
    pub fn success(data: T) -> Response<T> {
        Response::Success {
            result: "success".to_string(),
            data,
        }
    }

    pub fn failure(message: String) -> Response<T> {
        Response::Failure {
            result: "failure".to_string(),
            message,
        }
    }
}

pub fn success<T>(data: T) -> Response<T> {
    Response::success(data)
}

pub fn failure<T>(message: String) -> Response<T> {
    Response::failure(message)
}
